aiida_diff/parsers.py

```diff
@@ -42,29 +40,19 @@ def parse(self, **kwargs):
         """
         from aiida.orm import SinglefileData
 
-        # Check that the retrieved folder is there
-        try:
-            output_folder = self.retrieved
-        except exceptions.NotExistent:
-            return self.exit_codes.ERROR_NO_RETRIEVED_FOLDER
-
-        # Check the folder content is as expected
-        list_of_files = output_folder.list_object_names()
-        output_files = [self.node.get_option('output_filename')]
-        output_links = ['diff']
-        # Note: set(A) <= set(B) checks whether A is a subset
-        if set(output_files) <= set(list_of_files):
-            pass
-        else:
-            self.logger.error(
-                "Not all expected output files {} were found".format(
-                    output_files))
+        output_filename = self.node.get_option('output_filename')
 
-        # Use something like this to loop over multiple output files
-        for fname, link in zip(output_files, output_links):
+        # Check that folder content is as expected
+        files_retrieved = self.retrieved.list_object_names()
+        files_expected = [output_filename]
+        # Note: set(A) <= set(B) checks whether A is a subset of B
+        if not set(files_expected) <= set(files_retrieved):
```

Comment:
Maybe here you want to return an exit code. Or if you want to catch the case that you unexpectedly retrieved *more* files then keep the error. But then you should check explicitly for the presence of the `output_filename` because otherwise the following lines will except. I would say this check should just be replaced with a single explicit check for `output_filename` and otherwise return exit code

Reply:
> I would say this check should just be replaced with a single explicit check for output_filename 

It's of course a balancing act but here I think it is sensible to provide some code that can easily be extended to multiple output file names (which most codes have).

> and otherwise return exit code

Ok, where do we put the message that is currently in the exception?
Could you provide the line I should put?

Reply:
which exception? You can keep the logger error message which will end up in the report when you do `verdi calcjob report`, but then straight the line under I would return `self.exit_codes.ERROR_MISSING_OUTPUT_FILES` and then create that exit code on the `DiffCalculation` spec.

Reply:
> which exception?

Oh, I completely missed the transformation from exception to logger message.
Indeed, we need to return an exit code here.

Reply:
Coming back to this: One thing I haven't really figured out is when to use exit codes and when to use exceptions.

Is there a good reason why the parser should never except?
It's just that exit codes don't provide any flexibility in customizing the error message depending on the specific instance of the exception, so isn't it preferable to except here?

Reply:
The reason that I don't think one should use exceptions is that it will be impossible for the caller (note caller here is another process, not a human) to determine what went wrong and how to react, because it will have to "parse" the human readable message. The exit code is the best way for callers to base their response on. If you *also* want to give information to the human, then the logger is the perfect way. Just like in the `WorkChain` this is intended to be read by humans. All those logging statements will end up in `verdi process report`.

There is the option to maybe make the exit code message somewhat configurable when you retrieve it from the spec. I already opened an [issue for this sometime ago for discussion](https://github.com/aiidateam/aiida_core/issues/2644). However, I think we should still stress that exit codes should be seen exclusively to be used programmatically. Any messages that are intended for humans should be done as log messages.

Reply:
As an addendum: exceptions should never be explicitly used, also not in the parser, because an excepted process means something unexpected happened, such as a real error in the parser code. If one also starts to use it to just communicate failure, we lose all consistency

Reply:
> it will be impossible for the caller (note caller here is another process, not a human) to determine what went wrong and how to react, because it will have to "parse" the human readable message

Well, python has exception classes for this purpose, and the caller could simply check what type of exception was thrown. Is there a reason one should not do this, or is it simply not how our processes are currently implemented?

(by the way, this is just for my understanding - I'll put the exit code for the time being)

Reply:
You cannot catch an exception from a sub process in a process, because you are not actually running it. The engine is doing this in its event loop. Its just not the way the processes are implemented. It really resembles the implementation of UNIX processes, they also communicate to callers through an exit code

Reply:
alright, thanks!